runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/InMemoryJobService.java

```diff
@@ -161,6 +182,26 @@ public void run(RunJobRequest request, StreamObserver<RunJobResponse> responseOb
           invoker.invoke(
               preparation.pipeline(), preparation.options(), request.getRetrievalToken());
       String invocationId = invocation.getId();
+
+      if (cleanArtifactsPerJob) {
+        invocation.addStateListener(
+            state -> {
+              if (
+              // TODO: add other terminal states
+              state.equals(JobState.Enum.DONE) || state.equals(JobState.Enum.FAILED)) {
+                String stagingSessionToken = stagingSessionTokens.get(preparationId);
+                try {
+                  cleanupJobFn.accept(stagingSessionToken);
```

Comment:
Given its job Termination/Cleanup listener, we should be passing the job id or job object.